rust-symcrypt/src/hmac.rs

```diff
@@ -84,6 +110,372 @@ pub trait HmacState: Clone {
     fn result(self) -> Self::Result;
 }
 
+/// `HmacMd5ExpandedKey` is a struct that represents the expanded key for the [`HmacMd5State`].
+#[cfg(feature = "md5")]
+struct HmacMd5ExpandedKey(symcrypt_sys::SYMCRYPT_HMAC_MD5_EXPANDED_KEY);
+// Wrapping the expanded key so that it can be independently dropped after it's ref count has gone to 0
+
+// Since HmacMd5ExpandedKey can be referenced multiple times, HmacMd5ExpandedKey must be ref counted and a there needs to be a separate drop()
+#[cfg(feature = "md5")]
+impl Drop for HmacMd5ExpandedKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.0) as *mut c_void,
+                mem::size_of_val(&mut self.0) as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+}
+
+/// `HmacMd5State` is a struct that represents a stateful Hmac using MD5 and implements the [`HmacState`] trait.
+#[cfg(feature = "md5")]
+pub struct HmacMd5State {
+    // SymCrypt expects the address for its structs to stay static through the struct's lifetime to guarantee that structs are not memcpy'd as
+    // doing so would lead to use-after-free and inconsistent states.
+
+    // Using an `HmacMd5Inner` state that is Pin<Box<>>'d. Memory allocation is not handled by SymCrypt and Self is moved
+    // around when returning from `HmacMd5State::new()`. Box<> heap allocates the memory and ensures that it does not move
+    // within its lifetime.
+    inner: Pin<Box<HmacMd5Inner>>,
+}
+
+#[cfg(feature = "md5")]
+struct HmacMd5Inner {
+    state: symcrypt_sys::SYMCRYPT_HMAC_MD5_STATE,
+
+    // Must Arc<> the expanded_key field since it must be ref counted, clones of HmacSha265State will reference the same expanded key. Pin<> on the key
+    // is to ensure that address is not moved throughout the expanded key's lifetime.
+    //
+    // This semantic is not needed for the state field since it is initialized in line with HmacMd5Inner initialization.
+    expanded_key: Pin<Arc<HmacMd5ExpandedKey>>,
+}
+
+#[cfg(feature = "md5")]
+unsafe impl Send for HmacMd5Inner {
+    // Md5 Send/Sync is not needed for rustls, adding for consistency.
```

Comment:
>// Md5 Send/Sync is not needed for rustls, adding for consistency. [](http://example.com/codeflow?start=4&length=66)

It doesn't really make sense to use a single hash object across multiple threads concurrently. By implementing this, doesn't it imply that this struct is thread-safe? But SYMCRYPT_HMAC_MD5_STATE is *not* thread safe.

Reply:
Spoke offline. New github issue filed, and we will remove the `Send/Sync` traits for algos not required by rustls

Reply:
Removing these Send/Sync functions for the ones not needed for rustls, filed github issue.